content/300-guides/150-testing/100-unit-testing.mdx

```diff
@@ -0,0 +1,311 @@
+
+- Quickly find and isolate bugs in code.
+- Provides documentation for each module of code by way of indicating what certain code blocks should be doing.
+- A helpful gauge that a refactor has gone well. The tests should still pass after code has been refactored.
+
+In the context of Prisma, this generally means testing a function which makes database calls using the Prisma client.
+
+A single test should focus on the data returned by the Prisma client and how the function handles that data, ensuring that it behaves as expected.
+
+This means that you should aim to remove as many dependencies as possible, such as external services and databases, to keep the tests and their environments as lightweight as possible.
+
+</TopBlock>
+
+#### Prerequisites
+
+This guide assumes you have the JavaScript testing library [`Jest`](https://jestjs.io/) already setup in your project. If you do not then the following short guide will help you to get up and running.
+
+<details id="add-jest-to-project" open><summary>Add Jest to your project</summary>
+
+Prisma uses TypeScript and Jest can be setup to use TypeScript with the help of the [`ts-jest`](https://github.com/kulshekhar/ts-jest) package.
+
+```terminal
+npm install --save-dev jest typescript ts-jest @types/jest
+```
+
+Next create a `jest.config.js` file at your projects root, adding the following configuration:
+
+```js file=jest.config.js
+module.exports = {
+  clearMocks: true,
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+  setupFilesAfterEnv: ['<rootDir>/singleton.ts'],
+}
+```
+
+The above configuration declares a path to a `singleton.ts` file, this can be called anything you like, the following guide will name it `singleton.ts`. The module the path points towards is the code that runs setting up the testing framework before each test file is executed.
+
+Add a script to your `package.json` file which will run the tests. The `-i` flag will ensure that the tests are run one at a time to avoid any race conditions. The scripts will look for a folder at your projects root called `__tests__` (double\_\_underscore), and run any tests within files, within that folder.
+
+```json file=package.json
+  "scripts": {
+    ...otherScripts
+    "test": "jest -i"
+  },
+```
+
+</details>
+
+## Mocking the Prisma client
+
+To ensure your unit tests are isolated from external factors you can mock the Prisma client, this means you get the benefits of being able to use your schema (**_type-safety_**), without having to make actual calls to your database when your tests are run.
+
+This guide will cover two approaches to mocking the client, a singleton instance and dependency injection. Both have their merits depending on your use cases. To help with mocking the client the [`jest-mock-extended`](https://github.com/marchaos/jest-mock-extended) package will be used.
+
+```terminal
+npm install jest-mock-extended --save-dev
+```
+
+### Singleton
```

Comment:
If this is part of a guide I would call it 'Create a singleton', and the next section 'Inject the mock context' or whatever makes sense, so that the heading are 'do stuff'